### PR TITLE
GEODE-4101: Add --redirect-output flag to GFSH

### PR DESCRIPTION
**Pull request by jujoramos**
To avoid major code changes the start commands check the new flag and
set as `true` a new system property, `DISABLE_REDIRECTION_CONFIGURATION`,
which disables the other two (now deprecated) properties. In the next
major release, the properties and the related logic should be deleted.
The integration between log4j2 and geode-pulse was also incomplete.
When ran in embedded mode, some messages were totally swallowed and the
actual logging was split between the member's log and pulse.log file.

- Added flag `--redirect-ouput` to `StartServerCommand`.
- Added flag `--redirect-ouput` to `StartLocatorCommand`.
- Added tests to verify the behavior of the commands/builders.
- Added `OSProcess.DISABLE_REDIRECTION_CONFIGURATION` internal property.
- Property `OSProcess.ENABLE_OUTPUT_REDIRECTION` marked as deprecated.
- Property `OSProcess.DISABLE_OUTPUT_REDIRECTION` marked as deprecated.
- Added `log4j-jcl` as a dependency of `geode-pulse`. By adding this
  dependency no logging is lost and nothing is split: everything goes
  to the pulse.log file (or whatever is configured in `log4j2.xml`).

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by jdeppe-pivotal** [comment]
Thanks for the PR @jujoramos! Unfortunately it's failing the following test: `org.apache.geode.management.internal.cli.GfshParserAutoCompletionTest`. Please could you take a look at that?

File: geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartServerCommandTest.java [hunk 1]
```diff
@@ -141,4 +147,97 @@ public void testCreateServerCommandLineWithRestAPI() throws Exception {
...
+    expectedCommandLineElements.add("--max-connections=800");
+    expectedCommandLineElements.add("--max-message-count=500");
+    expectedCommandLineElements.add("--max-threads=100");
+    expectedCommandLineElements.add("--message-time-to-live=93");
+    expectedCommandLineElements.add("--socket-buffer-size=1048576");
+    expectedCommandLineElements.add("--hostname-for-clients=localhost");
+
+    assertNotNull(commandLineElements);
+    assertTrue(commandLineElements.length > 0);
+    assertEquals(commandLineElements.length, expectedCommandLineElements.size());
```
Comment (pdxrunner) [inline geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartServerCommandTest.java]: Delete this assertion `expectedCommandLineElements` only contains a subset of `commandLineElements`.

File: geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartLocatorCommandTest.java [hunk 2]
```diff
@@ -144,4 +149,110 @@ private void addJvmOptionsForOutOfMemoryErrors(final List<String> commandLine) {
...
+    for (String commandLineElement : commandLineElements) {
+      expectedCommandLineElements.remove(commandLineElement);
+    }
+
+    assertTrue(String.format("Expected ([]); but was (%1$s)", expectedCommandLineElements),
+        expectedCommandLineElements.isEmpty());
+  }
+
+  @Test
+  public void testCreateStartLocatorCommandLineWithAllOptions() throws Exception {
```
Comment (pdxrunner) [inline geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartLocatorCommandTest.java]: My comments  that follow on StartServerCommandTest apply to this test class as well.

**Comment by jujoramos** [comment]
Thanks both for reviewing!.
@jdeppe-pivotal: I just checked and, as you stated, the test `GfshParserAutoCompletionTest` is failing. I'll check the test and fix the problems but, considering that we have continuous-integration enabled, shouldn't it have failed when I opened the pull request?. Where did you see the failure (just for future reference so I can fix these things before the actual review)?.
@pdxrunner: I'll make the modifications and commit the changes, thanks!.

File: geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartLocatorCommandTest.java (hunk 2 above)
Comment (jujoramos) [reply geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartLocatorCommandTest.java]: @pdxrunner: same as below, the `expectedCommandLineElements` variable should contain the full set of expected elements:

```
    Set<String> expectedCommandLineElements = new HashSet<>();
    expectedCommandLineElements.add(StartMemberUtils.getJavaPath());
    expectedCommandLineElements.add("-server");
    expectedCommandLineElements.add("-classpath");
    expectedCommandLineElements
        .add(StartMemberUtils.getGemFireJarPath().concat(File.pathSeparator).concat(customClasspath)
            .concat(File.pathSeparator).concat(StartMemberUtils.CORE_DEPENDENCIES_JAR_PATHNAME));
    expectedCommandLineElements
        .add("-DgemfirePropertyFile=".concat(gemfirePropertiesFile.getAbsolutePath()));
    expectedCommandLineElements.add(
        "-DgemfireSecurityPropertyFile=".concat(gemfireSecurityPropertiesFile.getAbsolutePath()));
    expectedCommandLineElements.add("-Dgemfire.statistic-sample-rate=1500");
    expectedCommandLineElements.add("-Dgemfire.disable-auto-reconnect=true");
    expectedCommandLineElements.addAll(Arrays.asList(jvmArguments));
    expectedCommandLineElements.add("-Xms".concat(heapSize));
    expectedCommandLineElements.add("-Xmx".concat(heapSize));
    expectedCommandLineElements.add("-XX:+UseConcMarkSweepGC");
    expectedCommandLineElements.add("-XX:CMSInitiatingOccupancyFraction="
        .concat(String.valueOf(StartMemberUtils.CMS_INITIAL_OCCUPANCY_FRACTION)));
    expectedCommandLineElements.add("-Dgemfire.launcher.registerSignalHandlers=true");
    expectedCommandLineElements.add("-Djava.awt.headless=true");
    expectedCommandLineElements.add("-Dsun.rmi.dgc.server.gcInterval=9223372036854775806");
    expectedCommandLineElements.add("-Dgemfire.OSProcess.DISABLE_REDIRECTION_CONFIGURATION=true");
    expectedCommandLineElements.add("org.apache.geode.distributed.LocatorLauncher");
    expectedCommandLineElements.add("start");
    expectedCommandLineElements.add("customLocator");
    expectedCommandLineElements.add("--debug");
    expectedCommandLineElements.add("--force");
    expectedCommandLineElements.add("--hostname-for-clients=localhost");
    expectedCommandLineElements.add("--port=10101");
    expectedCommandLineElements.add("--redirect-output");
```

File: geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartServerCommandTest.java (hunk 1 above)
Comment (jujoramos) [reply geode-assembly/src/test/java/org/apache/geode/management/internal/cli/commands/StartServerCommandTest.java]: @pdxrunner: this is not accurate, `expectedCommandLineElements` contains **the full set of expected elements** in `commandLineElements`:

```
    Set<String> expectedCommandLineElements = new HashSet<>();
    expectedCommandLineElements.add(StartMemberUtils.getJavaPath());
    expectedCommandLineElements.add("-server");
    expectedCommandLineElements.add("-classpath");
    expectedCommandLineElements
        .add(StartMemberUtils.getGemFireJarPath().concat(File.pathSeparator).concat(customClasspath)
            .concat(File.pathSeparator).concat(StartMemberUtils.CORE_DEPENDENCIES_JAR_PATHNAME));
    expectedCommandLineElements
        .add("-DgemfirePropertyFile=".concat(gemfirePropertiesFile.getAbsolutePath()));
    expectedCommandLineElements.add(
        "-DgemfireSecurityPropertyFile=".concat(gemfireSecurityPropertiesFile.getAbsolutePath()));
    expectedCommandLineElements.add("-Dgemfire.statistic-sample-rate=1500");
    expectedCommandLineElements.add("-Dgemfire.disable-auto-reconnect=true");
    expectedCommandLineElements.addAll(Arrays.asList(jvmArguments));
    expectedCommandLineElements.add("-XX:OnOutOfMemoryError=kill -KILL %p");
    expectedCommandLineElements.add("-Xms".concat(heapSize));
    expectedCommandLineElements.add("-Xmx".concat(heapSize));
    expectedCommandLineElements.add("-XX:+UseConcMarkSweepGC");
    expectedCommandLineElements.add("-XX:CMSInitiatingOccupancyFraction="
        .concat(String.valueOf(StartMemberUtils.CMS_INITIAL_OCCUPANCY_FRACTION)));
    expectedCommandLineElements.add("-Dgemfire.launcher.registerSignalHandlers=true");
    expectedCommandLineElements.add("-Djava.awt.headless=true");
    expectedCommandLineElements.add("-Dsun.rmi.dgc.server.gcInterval=9223372036854775806");
    expectedCommandLineElements.add("-Dgemfire.OSProcess.DISABLE_REDIRECTION_CONFIGURATION=true");
    expectedCommandLineElements.add("org.apache.geode.distributed.ServerLauncher");
    expectedCommandLineElements.add("start");
    expectedCommandLineElements.add("fullServer");
    expectedCommandLineElements.add("--assign-buckets");
    expectedCommandLineElements.add("--debug");
    expectedCommandLineElements.add("--disable-default-server");
    expectedCommandLineElements.add("--force");
    expectedCommandLineElements.add("--rebalance");
    expectedCommandLineElements.add("--redirect-output");
    expectedCommandLineElements.add("--server-port=41214");
    expectedCommandLineElements.add("--spring-xml-location=/config/spring-server.xml");
    expectedCommandLineElements.add("--critical-heap-percentage=95.5");
    expectedCommandLineElements.add("--eviction-heap-percentage=85.0");
    expectedCommandLineElements.add("--critical-off-heap-percentage=95.5");
    expectedCommandLineElements.add("--eviction-off-heap-percentage=85.0");
    expectedCommandLineElements.add("--max-connections=800");
    expectedCommandLineElements.add("--max-message-count=500");
    expectedCommandLineElements.add("--max-threads=100");
    expectedCommandLineElements.add("--message-time-to-live=93");
    expectedCommandLineElements.add("--socket-buffer-size=1048576");
    expectedCommandLineElements.add("--hostname-for-clients=localhost");
```

**Comment by jujoramos** [comment]
@pdxrunner / @jdeppe-pivotal: is there anything else you want to change here?, or can we merge this one already?.

**Comment by pdxrunner** [comment]
LGTM. My precheckin run just finished green

**Pull request merged**